codex/stores/maintenance.nim

```diff
@@ -10,92 +10,323 @@
+  ##
+
+  trace "Updating a dataset expiry", treeCid, minExpiry
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
+
+  await modify[DatasetMetadata](self.metaDs, key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        let datasetMd = DatasetMetadata(
+          expiry: max(currDatasetMd.expiry, minExpiry),
+          leavesCount: currDatasetMd.leavesCount,
+          manifestsCids: currDatasetMd.manifestsCids,
+          checkpoint: currDatasetMd.checkpoint
+        )
+        return datasetMd.some
+      else:
+        raise newException(CatchableError, "DatasetMetadata for treeCid " & $treeCid & " not found")
+  )
+
+
+proc recordCheckpoint*(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  # Saves progress or deletes dataset metadata if progress > leavesCount
+  #
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
 
-proc runBlockCheck(self: BlockMaintainer): Future[void] {.async.} =
-  let expirations = await self.repoStore.getBlockExpirations(
-    maxNumber = self.numberOfBlocksPerInterval,
-    offset = self.offset
+  await self.metaDs.modify(key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        if currDatasetMd.expiry != datasetMd.expiry or currDatasetMd.manifestsCids != datasetMd.manifestsCids:
+          raise newException(CatchableError, "Change in expiry detected, interrupting maintenance for dataset with treeCid " & $treeCid)
+
+        if currDatasetMd.checkpoint.progress > datasetMd.checkpoint.progress:
+          raise newException(CatchableError, "Progress should be increasing only, treeCid " & $treeCid)
+
+        if currDatasetMd.leavesCount <= datasetMd.checkpoint.progress:
+          DatasetMetadata.none
+        else:
+          datasetMd.some
+      else:
+        raise newException(CatchableError, "Metadata for dataset with treeCid " & $treeCid & " not found")
   )
 
-  without iter =? expirations, err:
-    trace "Unable to obtain blockExpirations iterator from repoStore"
-    return
+proc deleteBatch(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  var datasetMd = datasetMd
+
+  datasetMd.checkpoint.timestamp = self.clock.now
+
+  if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+    return failure(err)
+
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    startingIndex = datasetMd.checkpoint.progress
+
+  trace "Deleting a batch of blocks", size = self.batchSize
+
+  var index = datasetMd.checkpoint.progress
+  while (index < datasetMd.checkpoint.progress + self.batchSize) and
+          (index < datasetMd.leavesCount):
+    if err =? (await self.blockStore.delBlock(treeCid, index)).errorOption:
```

Comment:
We should probably have the ability to delete batches of blocks, instead of issuing a request per block. We have batching primitives in the datastore, but they aren't exposed in the datastore interface, which we should probably do.

Reply:
Ok, please create an issue in `nim-datastore` for this. If there will be some concurrent safe API for handling batch updates I will happily use it.